src/module_4/module_4_model-fitting.md

```diff
@@ -0,0 +1,1214 @@
+    
+
+
+### Model 2: Random Forest - Without categorical variables - variables with importance > 0.02
+
+La distribución de la importancia de los coeficientes en nuestro modelo no parece equilibrada. Por tanto, procederemos a desarrollar un nuevo modelo simplificado, descartando aquellos coeficientes que presenten una importancia inferior a 0.02.
+
+
+```python
+train_cols = ["global_popularity", "days_since_purchase_product_type", "user_order_seq", "days_since_purchase_variant_id", "abandoned_before", "std_days_to_buy_variant_id", "avg_days_to_buy_variant_id", "discount_pct", "normalised_price", "ordered_before", "std_days_to_buy_product_type"]
+
+n_trees = 100
+
+rf_2 = RandomForestClassifier(n_estimators=n_trees)
+rf_2.fit(X_train[train_cols], y_train)
+    
+train_probs = rf_2.predict_proba(X_train[train_cols])[:, 1]
+test_probs = rf_2.predict_proba(X_test[train_cols])[:, 1]
+val_probs = rf_2.predict_proba(X_val[train_cols])[:, 1]
+
+print(f"Train_auc: {roc_auc_score(y_train, train_probs)}.",f"Val_auc {roc_auc_score(y_val, val_probs)}")
+print(f"Train_ap {average_precision_score(y_train, train_probs)}.",f"Val_ap {average_precision_score(y_val, val_probs)}")
+print(f"Train_ce {log_loss(y_train, train_probs)}.",f"Val_ce {log_loss(y_val, val_probs)}")
+
+plot_metrics("Random_Forest_2", test_probs, y_test)
+plot_metrics("Random_Forest_2", val_probs, y_val)
+```
+
+    Train_auc: 0.9951735546753651
+    Val_auc 0.7437482863079924
+    Train_ap 0.8479491372821064
+    Val_ap 0.110158456175734
+    Train_ce 0.021404713671575235
+    Val_ce 0.18508661408635912
+
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_21_1.png)
+    
+
+
+    Precision Recall AUC: 0.11948090674617384
+    Roc Curve AUC: 0.7434437365377136
+
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_21_3.png)
+    
+
+
+    Precision Recall AUC: 0.11331897652079789
+    Roc Curve AUC: 0.7437482863079924
+
+
+
+```python
+plot_feature_importance(rf_2, train_cols)
+```
+
+
+
+
+    (<Figure size 640x480 with 1 Axes>, <Axes: ylabel='features'>)
+
+
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_22_1.png)
+    
+
+
+Parece que, en este modelo específico, la eliminación de los coeficientes con menor importancia no ha tenido un impacto significativo.
+
+## Gradient Boosting Trees
+
+Para continuar avanzando con la práctica, experimentaremos con modelos de Gradient Boosting Trees, utilizando los modelos anteriores como punto de referencia. Comenzaremos ajustando los distintos parámetros de nuestro modelo para identificar la configuración más óptima. Para simplificar el proceso, fijaremos el número de árboles en 100 como valor predeterminado para el modelo, ya que realizar iteraciones como en ocasiones anteriores consume demasiado tiempo.
+
+
+```python
+train_cols = numerical_cols + binary_cols
+
+learning_rates = [0.05, 0.1]
+max_depths = [1, 3, 5]
+n_trees_grid = [100] 
```

Comment:
Bear in mind that in order to train with 100 trees, you first need to train with 1, 2, 3... until 100. This means that you can know your performance for any number of trees up to 100. Can you use this to know what is the optimal number of trees that you need to use? Plot learning curves